sklearn/linear_model/_coordinate_descent.py

```diff
@@ -816,6 +816,12 @@ class ElasticNet(MultiOutputMixin, RegressorMixin, LinearModel):
     To avoid unnecessary memory duplication the X argument of the fit method
     should be directly passed as a Fortran-contiguous numpy array.
 
+    The precise stopping criteria based on `tol` are the following: First, check that
+    that maximum coordinate update, i.e. :math:`\\max_j |w_j^{new} - w_j^{old}|`
+    is smaller than `tol` times the maximum absolute coefficient, :math:`\\max_j |w_j|`.
+    If so, then additionally check whether the dual gap is smaller than `tol` times
+    :math:`||y||_2^2 / n_{\text{samples}}`.
```

Comment:
Is this `||y||_2^2` part describing the following code?

https://github.com/scikit-learn/scikit-learn/blob/59428da95751fa92d46687850932f8f6ab1b4f3d/sklearn/linear_model/_cd_fast.pyx#L238

I do not see where `n_samples` come into this? (I could be missing something)

Reply:
We use different formulations in the python code like `ElasticNet` and the cython code in `_cd_fast.pyx`. Therefore, we divide the dual gap by `1/n_samples`, see https://github.com/scikit-learn/scikit-learn/blob/59428da95751fa92d46687850932f8f6ab1b4f3d/sklearn/linear_model/_coordinate_descent.py#L658-L660

Reply:
The `||y||_2^2` part comes from here https://github.com/scikit-learn/scikit-learn/blob/59428da95751fa92d46687850932f8f6ab1b4f3d/sklearn/linear_model/_cd_fast.pyx#L158-L159